Change mentions of optional null safety (#4819)
dart-lang/site-www: src/codelabs/null-safety.md

Sometimes the flow of the program tells you that the value of an
expression cannot be `null`.
To force Dart to treat that expression as non-nullable,
add the [null assertion operator](#the-null-assertion-operator-) (`!`).
If the value does equal `null`, using this operator throws an exception.

To handle potential `null` values, use the conditional property access
operator (`?.`) or null-coalescing operators (`??`)
to conditionally access a property or
provide a default value if `null` respectively.

### Exercise: Conditional property access

If you don't know that an expression with a nullable type equals `null` or not,
you can use the conditional member access operator (`?.`).
With this operator,
you can execute the remainder of the expression based on the condition that
`nullableObject` has a value.

<?code-excerpt "null_safety_codelab/bin/null_aware_operators.dart (conditional-property-access)"?>
```dart
// The following calls the 'action' method only if nullableObject is not null
nullableObject?.action();
```

In the following code, try using conditional property access
in the `stringLength` method. This fixes the error and

> to return the length of the string or `null` if it equals `null`:

Reviewer: Change "the error and to return the length" to "the error and returns the length" ?

## Type promotion

> Dart's [flow analysis](/null-safety/understanding-null-safety#flow-analysis)
> to account for nullability.

Reviewer 2: The grammar got messed up here.

```suggestion
Dart's [flow analysis](/null-safety/understanding-null-safety#flow-analysis)
accounts for nullability.
```

## Type promotion

> With sound null safety, we extended Dart's

Reviewer: I'd drop the history part of this (i.e. change "we extended Dart's flow analysis to account for..." to "Dart's flow analysis accounts for...")

### Exercise: Null checking

In the following code, add an `if` statement to the beginning of

> `getLength` that returns zero if `str` has no value:

Reviewer 2: I feel "is null" is better than "no value" because "no value" could also be taken to mean it hasn't been assigned a value yet.

```suggestion
`getLength` that returns zero if `str` is `null`:
```

[`late` keyword](/null-safety/understanding-null-safety#late-variables).

When you put `late` in front of a variable declaration,
that tells Dart the following about the variable:

> - It cannot be assigned a value yet.

Reviewer: This is too strong. I might be that it could be assigned a value, but that the code for that would just be too cumbersome.

`late` is the developer expressing *their* intent: "I don't want to assign assign value just yet"
